### ISSUE TITLE
Installed nested optional kernels silently fall back to the Torch implementation

### ISSUE THREAD

**Issue by DimensionSTP** [huggingface/transformers#48148]
### System Info

# Installed nested optional kernel silently falls back to Torch

## Summary

After the native-kernel fallback refactor in #47630, an installed optional backend can be ignored when its function is available only from a nested module and is not re-exported from the package root.

For Qwen3.5-family gated delta attention with `flash-linear-attention==0.5.0`, `fla.ops.gated_delta_rule.chunk_gated_delta_rule` imports directly, but `use_kernel_func_from_hub_with_fallback()` retains the Torch implementation. In our Qwen3.8-27B SFT workload this changed throughput from approximately 25-30 s/it to 169-200 s/it with the rest of the environment held fixed.

## Environment

- Transformers current `main`: `2b296305d7`
- First offending refactor: #47630 (`e91f7ef4d3`)
- `flash-linear-attention==0.5.0`

## Minimal reproduction

```python
from unittest.mock import patch
from transformers.integrations.hub_kernels import use_kernel_func_from_hub_with_fallback


def torch_fallback(*args, **kwargs):
    return "torch"


with patch(
    "transformers.integrations.hub_kernels.use_kernel_forward_from_hub",
    return_value=lambda function: function,
):
    wrapped = use_kernel_func_from_hub_with_fallback("chunk_gated_delta_rule", "fla")(torch_fallback)

implementation = next(cell.cell_contents for cell in wrapped.__closure__ if callable(cell.cell_contents))
print(implementation.__module__, implementation.__name__)
```

Current output:

```text
__main__ torch_fallback
```

Expected output:

```text
fla.ops.gated_delta_rule.chunk chunk_gated_delta_rule
```

## Root cause and proposed scope

`resolve_internal_import()` only checks attributes reachable from the imported package root. It does not import an unexposed nested module. If root resolution returns `None`, import `package.internal_path`, resolve `func_name` there, and otherwise preserve the existing Torch fallback.

A two-file minimal patch and regression test are ready. Is this generic resolver fix an acceptable direction for a PR?

### Who can help?

@vasqu @drbh 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

I observed this while training a Qwen3.5-family hybrid model with `flash-linear-attention==0.5.0`.

The same FLA installation is directly importable:

```python
from fla.ops.gated_delta_rule import chunk_gated_delta_rule

print(chunk_gated_delta_rule.__module__)
# fla.ops.gated_delta_rule.chunk
```

However, the generic optional-kernel fallback introduced through #47630 does not resolve this nested function and silently retains the Torch implementation.

This can be reproduced on the current Transformers main branch:

```python
from unittest.mock import patch

from transformers.integrations.hub_kernels import use_kernel_func_from_hub_with_fallback


def torch_fallback(*args, **kwargs):
    return "torch"


with patch(
    "transformers.integrations.hub_kernels.use_kernel_forward_from_hub",
    return_value=lambda function: function,
):
    wrapped = use_kernel_func_from_hub_with_fallback(
        "chunk_gated_delta_rule",
        "fla",
    )(torch_fallback)

implementation = next(
    cell.cell_contents
    for cell in wrapped.__closure__
    if callable(cell.cell_contents)
)

print(implementation.__module__, implementation.__name__)
```
Actual output:

```text
 __main__ torch_fallback
```

The package-root lookup performed by resolve_internal_import() cannot reach a nested module that is not re-exported by the package root.
The fallback path does not subsequently import fla.ops.gated_delta_rule, even though that module and function are available.

In an otherwise unchanged Qwen3.8-27B SFT environment, this changed training throughput from approximately 25-30 s/it to 169-200 s/it.

Tested conditions:

- affected Transformers version: 5.15.0 and current main
- direct nested FLA import: succeeds

### Expected behavior

When the optional backend is installed and the requested function is available from its configured nested module, the fallback resolver should select that implementation.

For this case, the selected function should be:

```text
fla.ops.gated_delta_rule.chunk.chunk_gated_delta_rule
```
 
The Torch implementation should be used only when the optional package or its configured nested module cannot be imported.

The existing priority should remain unchanged:

1. Hub kernel, when enabled and available
2. Installed optional backend
3. Torch fallback

**Comment by MannXo** [comment]
Can I take this? I have some time to work on it this week.

**Comment by Gambit-Checkmate** [comment]
Hi, I’d like to look into the optional-kernel fallback behavior here. I can first reproduce the silent Torch fallback with a nested optional kernel, then add a focused test that checks the expected warning or error before changing the loader path. Is there a preferred backend or kernel package to use for the reproduction?

**Comment by DimensionSTP** [comment]
Thanks both for offering to help. I reported this from a reproduced training regression and already have the minimal two-file resolver patch and focused regression test ready, as noted in the issue description, so I plan to open the PR shortly and link it here.

The concrete backend is `flash-linear-attention==0.5.0`, with `fla.ops.gated_delta_rule.chunk_gated_delta_rule` directly importable while the generic resolver silently retains the Torch fallback. The regression test can mock the import path, so it does not need FLA as a test dependency.

I would be happy to incorporate your review and any additional backend cases once the PR is open.

**Comment by DimensionSTP** [comment]
The fix is now available in #48221.

It includes the minimal nested optional-backend resolution change and a focused regression test. The full CI suite has passed with no failures.
Reviews and additional backend cases are welcome.

**Comment by DimensionSTP** [comment]
Closing this after retesting with the latest FLA packages.

In a clean environment with `flash-linear-attention==0.5.2` and `fla-core==0.5.2`, the current unmodified Transformers main correctly resolves:

`fla.ops.gated_delta_rule.chunk.chunk_gated_delta_rule`

I also verified the Qwen3.5 gated-delta CUDA forward and backward paths, including finite gradients. Under the same Transformers code, FLA 0.5.0 still selects the PyTorch fallback.

This confirms that the issue was caused by the older FLA package structure and is resolved by upgrading FLA rather than changing Transformers. The related PR #48221 has therefore been closed as well.

**Comment by DimensionSTP** [comment]
Reopening based on the additional evidence reported by @imvladikon. I reproduced the issue on the current Transformers `main` in an isolated `fla-core==0.5.2` environment without `flash-linear-attention`.

Environment:

- Transformers `main`: `05e078a1d2` (`5.16.0.dev0`)
- `fla-core==0.5.2`
- `flash-linear-attention`: not installed
- `fla.__path__` contained only the isolated core package
- `import fla` did not expose `fla.ops` or `fla.layers`

Results:

| Kernel | Existing resolver | Decorator selection | Direct nested import |
|---|---|---|---|
| `chunk_gated_delta_rule` | `None` | Torch fallback | succeeds |
| `chunk_kda` | `None` | Torch fallback | succeeds |

After directly importing the nested module, the existing resolver succeeds. This confirms that resolution currently depends on import order.

I also reproduced why the previous `0.5.2` verification passed: when both `fla-core` and `flash-linear-attention` are visible, importing `fla.layers` transitively attaches `fla.ops`, which masks the core-only failure.

Finally, I ran both directly imported FLA kernels on CUDA in the isolated core-only environment:

- `chunk_gated_delta_rule`: forward and backward passed; all input gradients were finite
- `chunk_kda`: forward and backward passed; all input gradients were finite
- peak reserved GPU memory: approximately 258 MB

Therefore, upgrading to FLA 0.5.2 alone does not resolve the issue for the supported core-only package layout.
